pulp_container/app/tasks/sync_stages.py

```diff
@@ -349,6 +379,117 @@ def create_blob(self, blob_data, deferred_download=True):
 
         return blob_dc
 
+    async def create_signatures(self, man_dc, signature_source):
+        """
+        Create signature declarative artifacts.
+
+        Signatures are currently supported only for image manifests and not manifest lists.
+
+        For sigstore, signatures are downloaded from a specified url. The number of signatures is
+        unknown upfront, need to download them in order by incrementing the index until hit 404.
+
+        Args:
+            man_dc: Declarative content instance of a related Manifest
+            signature_source: Source where to get signatures from
+        """
+
+        def _extract_data_from_signature(signature_raw):
+            gpg = gnupg.GPG()
+            crypt_obj = gpg.decrypt(signature_raw)
+            if not crypt_obj.data:
+                log.info(
+                    _(
+                        "It is not possible to read the signed document, GPG error: {}".format(
+                            crypt_obj.stderr
+                        )
+                    )
+                )
+                return
+
+            try:
+                sig_json = json.loads(crypt_obj.data)
+            except Exception as exc:
+                log.info(
+                    _(
+                        "Signed document cannot be parsed to create a signature for {}."
+                        " Error: {}".format(man_dc.content.digest, str(exc))
+                    )
+                )
+                return
+
+            # Currently, the only supported type is ATOMIC.
+            if sig_json["critical"]["type"] != SIGNATURE_TYPE.ATOMIC_FULL:
+                log.info(
+                    _(
+                        "Unsupported signature type {}, signature is not synced for {}.".format(
+                            signature_json["critical"]["type"], man_dc.content.digest
+                        )
+                    )
+                )
+                return
+
+            sig_json["signing_key_id"] = crypt_obj.key_id
+            sig_json["signature_timestamp"] = crypt_obj.timestamp
+            return sig_json
+
+        signature_dcs = []
+        if signature_source == SIGNATURE_SOURCE.SIGSTORE:
+            man_digest_reformatted = man_dc.content.digest.replace(":", "=")
+            # should the namespaced_upstream_name be used to grab a signature?
+            sig_relative_baseurl = f"{self.remote.upstream_name}@{man_digest_reformatted}"
+
+            async with aiohttp.ClientSession() as session:
```

Comment:
There are 2 places/cases where I need to download stuff and I can't use a downloader from a remote.
1. HEAD request to a public server (no auth configured on a remote should be used but I guess proxy configuration needs to be in place).
2. Here. Need to do multiple downloads from a public server (again no auth but proxy is needed). Number of downloads is not clear upfront. URL contains an index which needs to be incremented for each download attempt until you hit HTTP 404, then stop.

Any suggestions to handle this?
The goal is to reuse sessions and have cleaner code.

Custom downloaders for each case?
Do we want to extend plugin API to add some generic downloaders, where one can specify http method and have a session to reuse but without preset configuration from a remote?

@ipanova @mdellweg @lubosmj @dkliban @bmbouter

Reply:
I'm collecting greivances with the current downloader system on / attached to this issue: https://pulp.plan.io/issues/7352

I agree that the current download infrastructure is not ideal.

Reply:
I want to help, but I'm a little lost. Why in this case is the remote not the right way to generate the downloader?

Reply:
@bmbouter, because we are trying to download data, e.g., from an untrusted source. We may either get rid of some of the attributes (secrets) within the remote's downloader and renew them later or build a customized downloader that will have a separate session shared during the run.

Reply:
I see. Well in terms of helping you get where you need to go, if you don't want the downloader configured from the remote, you'll have to construct it some other way. A shared session would be good also +1 to that.

What about having a customized factory that produces these types of downloaders? That is kind of the intent of that object, to produce configured downloaders. Or do folks think we should use this challenge to guide architectural changes to the downloaders now? I'm inclined to fix this with what we already have in place, but if folks want to redesign now I'm ok with that too.

As an aside, I'm so confused. Why would we trust content from an untrusted source?

Reply:
I'll try to expand on what @lubosmj  said and what I originally posted and seem to fail to explain properly.

To my understanding, if I generate a downloader from a remote:
  - the session will [get all the tls configuration from it](https://github.com/pulp/pulpcore/blob/master/pulpcore/download/factory.py#L107-L126)
  - the session will [use auth and proxy configuration set on a remote](https://github.com/pulp/pulpcore/blob/master/pulpcore/download/http.py#L272)

Use case:
In addition to a usual source of content which remote is configured for, there is a separate resource for signatures for the content which remote refers it. Signatures are on a public server and do not need any auth to be provided.

More specific needs:
 - to make a HEAD HTTP request to the public server without any auth and use proxy if it's configured on a remote. (Not sure if a "downloader" name is suitable for this need since it's not a GET request)
 - make a many GET requests, N per each content, to the same public server with no auth and use proxy if it's configured on a remote.

Is it helpful or still confusing?

Reply:
> As an aside, I'm so confused. Why would we trust content from an untrusted source?

If it's a digital signature, it brings means of validation for itself, and there's no need to trust the source it was obtained from.

Reply:
Ah this makes sense re the signature server not needing to be trusted. Thanks.

Just to confirm, these signatures are only fetched at sync-time? Or are they fetched on-demand or streamed via the content app?

Reply:
After looking this over more with @goosemania I think two things:

1) For now, subclass the DownloaderFactory and use that to generate these downloaders. It will involve copying some code from the `_make_aiohttp_session_from_remote`method until we make that more granular.

2) @dralley what I think really needs improvement is the `DownloaderFactory`. We should find ways to subclass it with little to no code copying. To me that's at least one of the main issues.